### PR TITLE
Added base class for variational methods

### PR DESCRIPTION
**Pull request by ferrine**
Because of upcomming Normalizing flows and Stein variational gradient descent I propose a unified interface for variational methods

### CODE REVIEW

**Comment by twiecki** [comment]
Seems like you need to rebase.

**Comment by ferrine** [comment]
After some tests it will be ready for merge, so I need a review

File: pymc3/distributions/dist_math.py
Comment (springcoil) [inline pymc3/distributions/dist_math.py]: This is all clear and leverages theano in a good idiomatic way.

File: pymc3/math.py
Comment (springcoil) [inline pymc3/math.py]: Hmm using a private method that's the same name as the other one. I'm not so sure about this.

File: pymc3/variational/replacements.py
Comment (springcoil) [inline pymc3/variational/replacements.py]: I like this

Comment (springcoil) [reply pymc3/variational/replacements.py]: Good use of the theano.scan functionality, and hidden well in the API.

File: pymc3/math.py
Comment (springcoil) [reply pymc3/math.py]: Actually you've done it quite a lot - so this is just a style thing - I don't see a huge problem with it.

**Comment by springcoil** [comment]
Code looks clean - a few stylistic concerns but nothing major. @ferrine do you want another review before this merges?

File: pymc3/math.py
Comment (ferrine) [reply pymc3/math.py]: I was thinking about it too. It should be better named as `list_flatten` or something like this

**Comment by ferrine** [comment]
Thanks, @springcoil. Yes, I want some more reviews as I want Normalizing Flows and SVGD to be implemented through this unified api. CC @twiecki @taku-y

**Comment by ferrine** [comment]
What's up? `Cannot allocate memory`:(
https://travis-ci.org/pymc-devs/pymc3/jobs/184417786#L720

**Comment by twiecki** [comment]
Hm, that's odd. Have you tracked memory consumption locally? Could be a leak.

**Comment by ferrine** [comment]
Bottom neck of my implementation is `scan`. It is working for too long time on 10000 samples(maybe that use case is too rare to take care of) and I was returning all the samples instead of just mean. Any ideas of how to optimize advi?

**Comment by ferrine** [comment]
It is a good idea to refactor all this for working fast

File: pymc3/variational/replacements.py
Comment (taku-y) [inline pymc3/variational/replacements.py]: For supporting autoencoding VB with mini-batches, the terms of local and global RVs in `log_p_W` and `log_q_W` should be differently weighted. Specifically, the local RV terms in `log_p_W` and `log_q_W` should have the same weight as `log_p_D`.

Comment (ferrine) [reply pymc3/variational/replacements.py]: if you are about `pi` it is a trick for smoothing training process. If you set it close to zero you'll move to better point that maximises likelihood and then changing it to 1 you'll move to ELBO optimum.

Comment (twiecki) [reply pymc3/variational/replacements.py]: @ferrine Is there a reference for that trick?

Comment (ferrine) [reply pymc3/variational/replacements.py]: Ye, see [Blundell 2015](https://arxiv.org/pdf/1505.05424.pdf) *3.4. Minibatches and KL re-weighting* they perform such trick within epoch.

**Comment by ferrine** [comment]
I've decided to optimize code a lot, it performs really great: 2.3 sec for 10000 samples. I've totally changed API, thus I need a review.

**Comment by twiecki** [comment]
@taku-y Do you have time for another review? No worries if not.

**Review by springcoil — COMMENTED** [review]
LGTM, like these improvements.

**Comment by ferrine** [comment]
I think after some more reviews and suggestions it will be ready for merge.

**Comment by springcoil** [comment]
Yes maybe @taku-y @ColCarroll @aloctavodia @AustinRochford would be suitable people.

File: pymc3/variational/replacements.py
Comment (taku-y) [reply pymc3/variational/replacements.py]: In the reference, `pi` is defined for global RVs. But when using AEVB, both of `log_p_W` and `log_q_W` include global and local RVs. I thinks the local RV terms should not be weighted by `pi`. 

```
 +    @property
 +    def log_q_W_local(self):
 +        if self.local_dict['x']:
 +            x = flatten_list(self.local_dict['x'])
 +            mu = flatten_list(self.local_dict['means'].values())
 +            rho = flatten_list(self.local_dict['rhos'].values())
 +            _log_q_W_local_ = tt.sum(log_normal3(x, mu, rho))
 +            return _log_q_W_local_
 +        else:
 +            return 0
 +
 +    @property
 +    def log_q_W_global(self):
 +        raise NotImplementedError
 +
 +    @property
 +    def log_q_W(self):
 +        return self.log_q_W_global + self.log_q_W_local
```
The above code weights global and local RVs the same amount.

**Comment by ferrine** [comment]
That was easy and I'd better go to sleep)) 2:30am

**Review by springcoil — APPROVED** [review]

File: pymc3/step_methods/nuts.py
Comment (twiecki) [inline pymc3/step_methods/nuts.py]: Why is this file showing up as a diff?

Comment (ferrine) [reply pymc3/step_methods/nuts.py]: I have no idea

Comment (ColCarroll) [reply pymc3/step_methods/nuts.py]: probably because of my refactor of some of these --  `nuts`, `hmc`, `quad_potential` are all in a subfolder (`pymc3/step_methods/hmc/`) now, but perhaps you checked in an old copy?

**Comment by ferrine** [comment]
After implementation of path derivate I did not find any difference in convergence:) 
I used [bayesian_neural_network_advi.ipynb](https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/bayesian_neural_network_advi.ipynb) for that
**New** - advi with path trick
**New2** - advi without path trick
advi

**Comment by ferrine** [comment]
it works slower than the previous implementation also

**Comment by twiecki** [comment]
@ferrine That's disappointing. Just a single use-case though. In any case, we should keep both and a option to switch between them (with the default perhaps being the original ADVI).

File: pymc3/step_methods/nuts.py
Comment (twiecki) [reply pymc3/step_methods/nuts.py]: @ferrine Did you rebase to most recent `3.1`?

Comment (twiecki) [reply pymc3/step_methods/nuts.py]: I bet you accidentally checked in this file with a `git commit -a`, at least that happens to me all the time.

**Comment by ferrine** [comment]
I think that's because 2 shared variables instead of one

**Comment by ferrine** [comment]
One concatenated variable gives the same results

**Comment by ferrine** [comment]
Scan is the problem

**Comment by ferrine** [comment]
I've improved the performance and now advi is really fast even on 1+ samples:

1 sample ~ 950 it/sec (was 1000)
2 samples ~ 270 it/sec (was 7)

**Comment by ferrine** [comment]
For more than 1 samples we can unroll scan to get performance but it is not frequently used I think. Should I do optimisations there too?

Another point of optimization is here:
```python
    def posterior(self, initial):
        """Transforms initial latent space to posterior distribution

        Parameters
        ----------
        initial : initial latent space shape(samples, size)

        Returns
        -------
        Tensor
            posterior space
        """
        if self.local_vars:
            return tt.concatenate([
                self.posterior_local(initial[:, self.local_slc]),
                self.posterior_global(initial[:, self.global_slc])
                ], axis=1)
        else:
            return self.posterior_global(initial)
```
I thought that the first branch of IF statement is executed with C code but it appeared that it was pyhton. Execution time is 18% slower on neural network example without catching NoLocals case. 
All the rest code is running in C according to profiler

**Comment by twiecki** [comment]
> 1 sample ~ 950 it/sec (was 1000)

Seems like a slow-down for the 1-sample case?

> For more than 1 samples we can unroll scan to get performance but it is not frequently used I think. Should I do optimisations there too?

I don't think the > 1 samples case is important to optimize.

**Comment by ferrine** [comment]
I think that slow down is OK and worth it

**Comment by twiecki** [comment]
worth it for what? the multi sample case or the refactoring?

**Comment by ferrine** [comment]
All that

**Comment by ferrine** [comment]
It is minor slow down in 1 sample that gives lots of profits. I think old advi can be replaced in compatible way

**Comment by ferrine** [comment]
Do we need replacing old adv or just suggest a new way to do inference?

**Comment by twiecki** [comment]
I think we should probably replace it once we reach feature parity (mini-batch advi is still missing here I think), but will defer to @taku-y.

**Comment by ferrine** [comment]
@twiecki Minibatch "inference name" is supported out of the box as I do likelihood reweighting with given population size

File: pymc3/step_methods/nuts.py
Comment (twiecki) [reply pymc3/step_methods/nuts.py]: @ferrine Can you remove this file from the commit?

**Comment by twiecki** [comment]
@ferrine Cool, I didn't realize that. So mean-field here is ADVI?

Can you add an example of the new API with and without minibatch?

**Comment by ferrine** [comment]
Api is not ready yet. Previously I just optimized elbo by myself. There should be a more convenient way with generic interface. Maybe `pymc3.variational.inference` module will be that api.

**Comment by twiecki** [comment]
@ferrine OK, I think we should wait with merging this until the API is ready, do you agree?

**Comment by twiecki** [comment]
I'd be curious how SVGD would look with this.

**Comment by ferrine** [comment]
@twiecki Agree

File: pymc3/step_methods/nuts.py
Comment (ferrine) [reply pymc3/step_methods/nuts.py]: Just delete it? I was playing with rebasing for half an hour and ended up with `rm -rf pymc3; git clone ...`

**Comment by ferrine** [comment]
I've added a new generic function `approximate` it supports minibatches via `callbacks` and method specification for future improvements

Usage
------

```python
...
from pymc3.variational.replacements import MeanField
from pymc3.variational.inference import approximate
n = 1000                                                                
sd0 = 2.                                                                
mu0 = 4.                                                                
sd = 3.                                                                 
mu = -5.                                                                
                                                                        
data = sd * np.random.randn(n) + mu                                     
                                                                        
d = n / sd ** 2 + 1 / sd0 ** 2                                          
mu_post = (n * np.mean(data) / sd ** 2 + mu0 / sd0 ** 2) / d            
                                                                        
with Model():                                                           
    mu_ = Normal('mu', mu=mu0, sd=sd0, testval=0)                       
    Normal('x', mu=mu_, sd=sd, observed=data)                           
    mf = MeanField()                                                  
    approximate(30000, method=mf)
    # or mf, elbos = approximate(30000, method='advi')                            
    trace = mf.sample_vp(10000)                                         
```

**Comment by twiecki** [comment]
I like how that mirrors the sample() API. why not rename mean field to advi?

**Comment by ferrine** [comment]
Automatic Differentiation Variational Inference is badly coined name for algorithm, I used to name such approximation as mean field:) I agree with you that it is better to leave common, well known name instead. 
I also think that BaseReplacement is bad class name but can't imagine a better one. Do you have any ideas?

**Comment by taku-y** [comment]
I think "VariationalRepacement" may be better because it replaces RV nodes for variational approximation, although I also think "BaseReplacement" is OK if there is a docstring descrbing the objective :)

**Comment by ferrine** [comment]
PR seems to be ready for merge.

File: pymc3/variational/approximations.py
Comment (twiecki) [inline pymc3/variational/approximations.py]: Needs docs.

File: pymc3/variational/inference.py
Comment (twiecki) [inline pymc3/variational/inference.py:75]: Do all VI algos return `elbo_history`? What about SVGD?

File: pymc3/variational/inference.py
Comment (twiecki) [inline pymc3/variational/inference.py]: extra with?

File: pymc3/variational/approximations.py
Comment (twiecki) [reply pymc3/variational/approximations.py]: Pretty crazy that ADVI just becomes these few lines of code.

**Comment by twiecki** [comment]
@taku-y You're probably the most qualified to review this.

File: pymc3/variational/approximations.py
Comment (taku-y) [inline pymc3/variational/approximations.py:446]: This way of weighting is specialized to Blundell's paper (uncertainty of weights...). But it does not support other weighting scheme like autoencoding VB with global and local parameters. Since BaseApproximation is the base class for variational approximation, it would be better to make weighting be more flexible. This is the reason that I implemented advi_minibatch() such that each of the log probability terms can be weighted individually.

If the above statement is unclear, I can make another PR to remedy this issue after this PR is merged.

File: pymc3/variational/inference.py
Comment (taku-y) [reply pymc3/variational/inference.py:75]: I think SVGD can compute ELBO because we have samples from variational posterior. Is it wrong?

File: pymc3/variational/inference.py
Comment (ferrine) [reply pymc3/variational/inference.py]: You can define a callback that changes `pi` or number of samples depending on the train process. That can be useful for fine-tuning neural networks for example or sending emails about the current results.

File: pymc3/variational/inference.py
Comment (ferrine) [reply pymc3/variational/inference.py:75]: I've read the paper and SVGD seems to return something like samples from posterior. Am I right? If so, this class can not be used for SVGD easily

File: pymc3/variational/approximations.py
Comment (ferrine) [reply pymc3/variational/approximations.py:446]: Do you mean that with minibatch we underweight local vars because the evaluation is not on full dataset?

File: pymc3/variational/inference.py
Comment (ferrine) [reply pymc3/variational/inference.py]: `With` is about callbacks. Seems that I incorrectly understood you

File: pymc3/variational/approximations.py
Comment (taku-y) [reply pymc3/variational/approximations.py:446]: In the current implementation, KL_q_p_W = KL_q_p_W_local + KL_q_p_W_global. Only the global term should be wegited with pi and LK_q_p_W_local should be weighted with 1.0 as for the term log_p_D, because both of log_q_p_local and log_p_D are relevant to samples in mini-batches.

Comment (ferrine) [reply pymc3/variational/approximations.py:446]: Got it. So we have another problem: while likelihood is reweighted, local var associated with it is not. Thanks for detailed explanation

**Comment by twiecki** [comment]
I think it would be useful to have some examples to demonstrate the proposed api, specifically:
* Standard ADVI
* Mini-batch ADVI
* AEVB

**Comment by ferrine** [comment]
@twiecki, definitely, mnist might be a good example. But first I need to fix the reweighting problem with help of @taku-y.

**Comment by ferrine** [comment]
@taku-y, check please my fixes. I hope I've understood the problem

File: pymc3/variational/approximations.py
Comment (taku-y) [inline pymc3/variational/approximations.py]: @ferrine Sorry, my explanation was wrong. log_q_W_local should be weighted with (M * pi) where M is the number of mini-batches. According to Eq. (8) in the paper, the ELBO term relevant to samples should be scaled M times of  the global term. And prior KL terms are scaled proportional to pi, while leaving likelihood terms without scaling.

``` python
elbo = (self.log_p_D(posterior)
            - M * pi * self.log_q_W_local(posterior)
            - pi * self.log_q_W_global(posterior)
            + M * pi * self.log_p_W_local(posterior)
            + pi * self.log_p_W_global(posterior))
```

We can avoid hardcoding for weighting as follows:

``` python
elbo = (self.log_p_D(posterior)
            - pi_local * self.log_q_W_local(posterior)
            - pi_global * self.log_q_W_global(posterior)
            + pi_local * self.log_p_W_local(posterior)
            + pi_global * self.log_p_W_global(posterior))
```
In the later case, pi_local/global should be appropriately set by the user, or in a child class of BaseApproximation.

File: pymc3/variational/approximations.py
Comment (ferrine) [inline pymc3/variational/approximations.py:342]: @taku-y I do local reweighting here

File: pymc3/variational/approximations.py
Comment (ferrine) [reply pymc3/variational/approximations.py]: @taku-y thanks for pointing that out

File: pymc3/variational/approximations.py
Comment (taku-y) [reply pymc3/variational/approximations.py:342]: I see. So the ELBO terms should be weighted as

loglikelihood + M * pi * (log_p_global - log_q_global) + M * pi * (log_p_local - log_q_local)

where the sum of pi over all mini-batches is 1 as assumed in the paper. 

With pi=1/M, the sum of LHS of Eq. 9 (equal to eq.8) in the paper is equal to the ELBO, while the sum of the above objective equals to M times of ELBO. I prefer the later definition because at every step during optimization approximately true ELBO is obtained. 

Your previous code works with pi_new = M * pi (pi sum to 1).

Comment (taku-y) [reply pymc3/variational/approximations.py:342]: Though, the difference of the definitions of the objectives is only scaling.

**Comment by taku-y** [comment]
To make clear how weighting works in the current implementation, I wrote equations for each of terms in ELBO. You can see this with [Online Latex editor](https://www.codecogs.com/latex/eqneditor.php). 

T = \sum_{m=1}^{M}N_m = \textrm{Number of all samples} \\\\ \\\\
N_m = \textrm{Number of samples in {\it m}-th mini-batch} \\\\
M = \textrm{Number of mini-batches} \\\\
z_n^m = \textrm{{\it n}-th sample in {\it m}-th mini-batch} \\\\ \\\\
\sum_{m=1}^{M}\pi_m = 1 \\\\ \\\\
l_n^m = \log p(x|z_n^m),\hspace{0.1cm}k^g = \log p(w) - \log q(w),\hspace{0.1cm}k_n^m = \log p(z_n^m) - \log q(z_n^m) \\\\ \\\\
\sum_{m=1}^{M}\frac{T}{N_m}\sum_{n=1}^{N_m} l_n^m = T\sum_{m=1}^{M}\bar{l}^m=TM\bar{l} \\\\ \\\\
\sum_{m=1}^{M}M\pi_{m}k^g = Mk^g \\\\ \\\\
\sum_{m=1}^{M}M\pi_{m}\frac{T}{N_m}\sum_{n=1}^{N_m}k^m_n=TM\sum_{m=1}^{M}\pi_m\bar{k}^m = TM\bar{k}\\\\ \\\\

File: pymc3/distributions/dist_math.py
Comment (taku-y) [inline pymc3/distributions/dist_math.py:118]: Is this function used?

File: pymc3/distributions/dist_math.py
Comment (taku-y) [inline pymc3/distributions/dist_math.py]: What does "3" mean?

**Comment by taku-y** [comment]
Although I have minor comments, I think this PR is ready for merge, if @ferrine agree.

@ferrine You commented "preparing mnist vae" on a commit, are you planning to add MNIST VAE example in this PR?

**Comment by ferrine** [comment]
@taku-y, yes I'm preparing example of VAE using pymc3.1. It's rather new thing to me so it can take some time. By now I have notebook that ran on my laptop for a night but did not converge to a good optima. Planning to learn "how to aws" and experiment with gpu in some days

File: pymc3/distributions/dist_math.py
Comment (ferrine) [reply pymc3/distributions/dist_math.py]: this notion was used in other library:
 1) using sd
 2) using log sd
 3) using rho

File: pymc3/distributions/dist_math.py
Comment (ferrine) [reply pymc3/distributions/dist_math.py:118]: it is not used, I can delete it

File: pymc3/distributions/dist_math.py
Comment (fonnesbeck) [reply pymc3/distributions/dist_math.py]: Is there a more informative name for these functions that appending "3" to the name? Perhaps use `rho` in the name instead?

File: pymc3/variational/approximations.py
Comment (fonnesbeck) [inline pymc3/variational/approximations.py:52]: Need docstring here

File: pymc3/variational/approximations.py
Comment (fonnesbeck) [inline pymc3/variational/approximations.py]: Suggested edit:

> Approximation to posterior distribution using standard normal. Three methods are required for implementation:

File: pymc3/variational/approximations.py
Comment (fonnesbeck) [inline pymc3/variational/approximations.py]: New line after triple quote (throughout module), for consistency.

File: pymc3/variational/approximations.py
Comment (fonnesbeck) [reply pymc3/variational/approximations.py]: ADVI acronym ought to be all caps, just like MAP, etc.

Comment (ferrine) [reply pymc3/variational/approximations.py]: Agree

File: pymc3/variational/approximations.py
Comment (ferrine) [reply pymc3/variational/approximations.py:342]: In the current version of code you can set the population for local variable and just use weights that sum up to 1. You always know the population as it is equal to population of decoded observed variable. There is no need in `M` any more.

**Comment by ferrine** [comment]
I found that it is not possible to use theano type for shape of distribution. Can it be fixed?

File: pymc3/distributions/dist_math.py
Comment (ferrine) [reply pymc3/distributions/dist_math.py]: I think docstring with cross references will be better.

File: pymc3/variational/approximations.py
Comment (ferrine) [reply pymc3/variational/approximations.py:52]: Class docstring is always used to document init method(jupyter even has a popup help with shift+tab). All parent classes are supposed to be documented and thus will have docstrings for init. So I'll update a class docstring with signature description

**Comment by twiecki** [comment]
> I found that it is not possible to use theano type for shape of distribution. Can it be fixed?

No, @brandonwillard experimented with that but it's not easy.

Btw. does this work with updates.py from the svgd branch? Should we try and merge the two?

Also, what do we want to do with the current advi implementation? Deprecate it? Do we have feature parity now?

File: pymc3/distributions/dist_math.py
Comment (twiecki) [inline pymc3/distributions/dist_math.py]: These are just different parameterizations? Just giving them different numbers is a bit odd I think. Couldn't we merge them into a single function where you supply one of three kwargs: `std`, `w`, or `rho`?

File: pymc3/variational/inference.py
Comment (twiecki) [reply pymc3/variational/inference.py:75]: Yes, it does. Perhaps SVGD is too different, even though it's variational, it does not directly minimize the elbo via some approximation.

**Comment by taku-y** [comment]
@twiecki 

> Also, what do we want to do with the current advi implementation? Deprecate it? Do we have feature parity now?

I think it would be nice to leave the interface of the current advi functions unchanged for backward compatibility. On the other hand, these implementations should be refactored with the code of this PR for ease of maintainance.

**Comment by ferrine** [comment]
@twiecki 
I think warning message is enough until 3.2 then we can delete old advi. Code of old advi should use new implementation.

**Comment by twiecki** [comment]
Thanks @taku-y. I wonder if we should port advi() and advi_minibatch() to use this internally already (in this PR). It's really odd to have two implementations side-by-side. Would also make testing easier and have the examples be up-to-date. Shouldn't be too difficult I think, no?

**Comment by ferrine** [comment]
@twiecki refactoring old advi will be not difficult. I can include it in PR.

File: pymc3/variational/inference.py
Comment (ferrine) [reply pymc3/variational/inference.py:75]: [Operator Variational Inference](https://arxiv.org/abs/1610.09033) can be implemented with current API in my opinion. The main point is to distinguish elbo and objective in the implementation.

**Comment by twiecki** [comment]
that'd be great!

File: pymc3/distributions/dist_math.py
Comment (twiecki) [inline pymc3/distributions/dist_math.py:159]: Nice!

File: pymc3/variational/approximations.py
Comment (twiecki) [inline pymc3/variational/approximations.py]: *be

**Comment by ferrine** [comment]
After reading the paper about [Operator Variational Inference](https://arxiv.org/abs/1610.09033) I realised that all PR above can be done via OPVI. It opens the door to state of the art VI methods. I'm thinking about hard refactoring PR in order to handle general case of OPVI. 
I find it not very hard (but time consuming as I have some questions to terminology and methodology) since lots of the work devoted to standard ADVI can be reused during refactoring

**Comment by springcoil** [comment]
What are you proposing doing then Maxim? Refactoring this to use Operator VI?

**Comment by ferrine** [comment]
@springcoil exactly. Next(or even this one) PR will implement OPVI with LS objective that was described in the paper above

**Comment by springcoil** [comment]
Cool. I'd say split this into two PR's it just makes review so much easier :)

**Comment by ferrine** [comment]
That was pretty bad merge... So, I've tested scaling and we can ignore scaling densities in the future. This PR probably needs hard refactoring again.

**Comment by ferrine** [comment]
Hmm, I think we are able to add recompution of models logp so that it will be evaluated on a number of samples. This way we can avoid costly scan op when sampling logp. It will extend `Model`'s interface btw. I'll add it to PR

**Comment by ferrine** [comment]
I'm planning to refactor most of the code as I'm not satisfied with approach I use now. I'll do advi in OPVI framework soon. Implementing LG opvi will be in a couple of lines as well as advi. Here is a [draft](https://github.com/ferrine/pymc3/commit/64f1b2f51b453439490084f200e3deeedda154a6) for supposed opvi interface

**Comment by taku-y** [comment]
Sounds great. Hope your refactoring goes well. Btw, I'm curious about removing scan. Is it related to the shape of RV (#1125)?

**Comment by ferrine** [comment]
@taku-y, no, it's not related as I want to do elemwise computation on custom tensor. I'm afraid how it will look like on Multi Observed RV

**Comment by taku-y** [comment]
@ferrine Thanks for your reply. I shuold wait your code. I look forward to your new framework!

**Comment by ferrine** [comment]
I've tried to rebase the branch on master but it was full of pain. I would rather create a new branch and OPVI PR. I have lot's of things to refactor so there will be a mess soon otherwise

**Comment by ferrine** [comment]
Closing in favour of #1694

**Pull request closed without merging**
